[Merged by Bors] - feat(MeasureTheory/Integral): the Riesz-Markov-Kakutani theorem for `Real`-linear functionals

yoh-tanimoto opened the pull request:
Prove that `rieszContent` gives a measure `rieszMeasure` such that, under `Λ : C(X, ℝ) →ₗ[ℝ] ℝ) (hΛ : ∀ (f : C(X, ℝ)), 0 ≤ f → 0 ≤ Λ f)`, it holds that `∀ (f : C_c(X, ℝ≥0)), ∫ (x : X), f x ∂ (rieszMeasure Λ hΛ) = Λ f`. The measure is defined through `rieszContent` for `toNNRealLinear`-version of `Λ`. The result is first proved for `Real`-linear `Λ` because in a standard proof one has to prove the inequalities by considering `Λ f` and `Λ (-f)` for all functions `f`, yet on `C_c(X, ℝ≥0)` there is no negation.

---

Motivation : This will be a basis for example, of the spectral decomposition of self-adjoint operators on Hilbert spaces.

The `NNReal`-version is ~300 lines away.
https://github.com/leanprover-community/mathlib4/compare/yoh-tanimoto-linearRMK...yoh-tanimoto-NNRealRMK

- [x] depends on: #12266 for a variation of Urysohn's lemma.
- [x] depends on: #20040 for `rieszContentRegular`
- [x] depends on: #20257 for `toNNRealLinear`
- [x] depends on: #20258 for `MeasureTheory.integral_tsupport`
- [x] depends on: #21572 for renaming

sgouezel commented:
Instead of creating a new file, can you put your additions in the already existing file on the Riesz content? Otherwise, this makes the PR very hard to review (I don't know what is new material and what was already there).

yoh-tanimoto commented:
OK sorry, I wasn't sure whether I should change the existing file because I changed the whole setting from `ℝ≥0` to `ℝ`. Now I removed the new file and replaced the old file with mine.

sgouezel commented:
Oops, my bad, I hadn't noticed that you had switched from `NNReal` to `Real`. Can you explain the reason for this choice? It seems to me that most things down the road should be easier using `NNReal`-valued functions (in particular because the topology on spaces of measures is defined using such functions), so I'd rather use them unless they create a difficulty somewhere.

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -29,87 +34,425 @@ noncomputable section
+
+
+variable {X : Type*} [TopologicalSpace X] [LocallyCompactSpace X] [T2Space X] [NormalSpace X]
+variable (Λ : C(X, ℝ) →ₗ[ℝ] ℝ) (hΛ : ∀ (f : C(X, ℝ)), 0 ≤ f → 0 ≤ Λ f)
```
It's not reasonable to start with a linear form defined on all continuous functions, because in most applications you won't have this. For instance the approach to define Lebesgue measure using Riesz-Markov-Kakutani is first to define a linear form on continuous compactly supported functions using the classical Riemann integral, and then apply the theorem to get Lebesgue (note that we won't use this approach as we already have Lebesgue measure in mathlib, this is just for illustrative purposes). But Riemann integral doesn't define a linear form on all continuous functions, only on compactly supported ones!

j-loreaux replied:
Indeed, I agree with this, as there are *way* too many continuous functions on a locally compact space. @sgouezel we don't currently have a bundled version of the compactly supported continuous functions though, so what do you propose?

yoh-tanimoto replied:
yes, you are absolutely right! I wanted to change it to ` C₀(X, ℝ)` and forgot... I will try to do it. Or is it better to use compactly supported functions?

sgouezel replied:
One should indeed use compactly supported functions, because functions tending to zero at infinity are not always integrable -- so the Riemann integral for instance would not be well-defined on the function `1/(1+|x|)`, which tends to zero at infinity.

I think we should bundle compactly supported continuous functions to proceed there.

yoh-tanimoto replied:
the most general possible, sure!

I'm afraid I am not yet skilled enough to write new definitions, so it may take time to bundle compactly supported continuous functions. How should one proceed, is it a good idea to copy [ZeroAtInftyContinuousMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/ContinuousFunction/ZeroAtInfty.html#ZeroAtInftyContinuousMap), add say `has_compact_support'` and prove as many lemmas and coercions as possible?

sgouezel replied:
Exactly. You can copy the file to a new file `CompactlySupportedContinuousMap`, replace the field `zero_at_infty' with `HasCompactSupport toFun`, fix everything that breaks, and remove everything that doesn't make sense. This should be done in another PR, for easier reviewing. Don't hesitate to ask if you need help!

yoh-tanimoto replied:
I will try.

Should I replace `ZeroAtInfty` or extend it? I think it is convenient to have coercion from compactly supported functions to functions vanishing at infinity.

For the moment I use the symbol `C_c`, because apparently there is no small `c`. Is this fine?

sgouezel replied:
You should replace it. One can still register a coercion from compactly supported continuous functions to functions tending to zero at infinity even if it's not part of the definition -- this will require a proof, but that's normal. Otherwise, whenever someone defines a compactly supported functions, he would have to check both the compact support and the convergence to zero at infinity, which is a little bit silly.

yoh-tanimoto commented:
I changed from `NNReal` to `Real` because I have in mind the application of the RMK theorem to prove the spectral decomposition. There one would need even complex valued functionals, complex-valued measures, and so on, but I thought it was a good start to use linear structures rather than `NNReal`. What do you think?

sgouezel commented:
For complex measures, one should definitely use complex-valued functions. For real measures, real-valued functions. And for measures in the usual sense, which are nonnegative, then `NNReal`-valued functions looks like the most natural choice. We have already code duplication between measures and vector-valued measures, but this seems unavoidable.

sgouezel commented:
I mean for now we should just use `NNReal` and do everything with them. Once we have Riesz-Markov-Kakutani for measures, we may start thinking on what we want to do with real measures or complex measures (or vector-valued measures), but it wouldn't make sense to do it before we complete it for usual measures, because when doing it we will learn a lot.

yoh-tanimoto commented:
Ok, then I will bring back `NNReal` and the `Real` version as a separate file

sgouezel commented:
> Ok, then I will bring back `NNReal` and the `Real` version as a separate file

For now, I'd say we shouldn't even have the `Real` version at all, and only do the `NNReal` one.

yoh-tanimoto commented:
I see, I will do that.

yoh-tanimoto commented:
I tried to do the `NNReal` version myself, and I found that the proof  of additivity `rieszContentAux_eq_add` is painful: one should take, using Urysohn's lemma, a function `g` which is `0` on `K₁` and `1` on `K₂`, but then
- one has to prove that `g : X →ᵇ ℝ≥0`
- one has to prove that `1 - g : X →ᵇ ℝ≥0`
- `Mul` is not given for `X →ᵇ ℝ≥0`, so one has to prove that `f * (1-g) : X →ᵇ ℝ≥0`

I think this is what Kalle described [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/variations.20of.20Urysohn's.20lemma/near/434415114).

It seems that he and Jesse have the plan to do it, so I would leave this to them.

On the other hand, the `Real` version is done up to a variation of Urysohn's lemma. Let me know what I should do with this PR.

kkytola commented:
> I tried to do the `NNReal` version myself, and I found that the proof of additivity `rieszContentAux_eq_add` is painful: one should take, using Urysohn's lemma, a function `g` which is `0` on `K₁` and `1` on `K₂`, but then
> 
>     * one has to prove that `g : X →ᵇ ℝ≥0`
>     * one has to prove that `1 - g : X →ᵇ ℝ≥0`
>     * `Mul` is not given for `X →ᵇ ℝ≥0`, so one has to prove that `f * (1-g) : X →ᵇ ℝ≥0`
> 
> I think this is what Kalle described [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/variations.20of.20Urysohn's.20lemma/near/434415114).

This is indeed exactly the issue why our PRs stalled, and as I tried to quickly comment on in Zulip, explaning that we had discussed the necessary fixes for that with @ADedecker (before some related things were changed).

I think it is important to improve Mathlib so that `BoundedContinuousFunction` with values in `ℝ≥0` work reasonably (e.g. they have a multiplication and truncated subtraction). I think this fix should be done regardless of the exact implementation path we choose towards Riesz-Markov-Kakutani, and I will try to get started with that fix in the near future.

In an ideal mathlib-world, measures and "Lebesgue integration theory" (`lintegral`, i.e. integration of nonnegative functions) should play well with nonnegative bounded continuous functions. But I agree that in the current state of things, working with `NNReal` is making some things unnecessarily painful. Therefore I don't currently have a clear opinion on what is the most reasonable course of action for getting some implementation of Riesz-Markov-Kakutani to Mathlib.

yoh-tanimoto commented:
thanks for your comments. I think I understand the need to add more material about `BoundedContinuousFunction` as you say.

As for this PR, I defined `rieszContentNonneg`, which takes `RestrictNonneg Λ hΛ : C(X, ℝ≥0) → ℝ≥0`, so this essentially eqivalent to `rieszContentAux` in the file RieszMarkovKakutani. (I need to change `C(X, ℝ≥0)` to C_c(X, ℝ≥0), but waiting for the merge of #12402). I prove the additivity of it showing the additivity of the linear version of `rieszContentAux` and showing that they take the same values.

kkytola commented:
I made #12559 which enables subtraction for nonnegative bounded continuous functions. If the model (new type class and otherwise minimal changes) is considered reasonable, I will do a similar PR to allow also multiplication of nonnegative bounded continuous functions.

kkytola commented:
Now that #12790 and #12559 are in Mathlib and one can subtract and multiply bounded continuous functions, showing that the `rieszContentAux` gives rise to an actual content (`rieszContent` below) is simple. The following are translated steps from Jesse's Lean3 proof of this part, whose PRs unfortunately stalled exactly because subtraction and multiplication of bounded continuous functions needed to be done properly in Mathlib. (Jesse continued from here to construct the measure and characterize it.)

```
import Mathlib

open scoped Classical NNReal ENNReal BoundedContinuousFunction
open MeasureTheory Set Function

variable {Z : Type} [TopologicalSpace Z] [NormalSpace Z]

lemma exists_bounded_zero_one_of_closed_nnreal {Z : Type} [TopologicalSpace Z] [NormalSpace Z]
    {s t : Set Z} (s_closed : IsClosed s) (t_closed : IsClosed t) (disj : Disjoint s t) :
    ∃ (f : Z →ᵇ ℝ≥0), EqOn f 0 s ∧ EqOn f 1 t ∧ ⇑f ≤ 1 := by
  obtain ⟨g, g_zero_on_s, g_one_on_t, g_bdd⟩ := exists_bounded_zero_one_of_closed s_closed t_closed disj
  refine ⟨g.nnrealPart, ?_, ?_, ?_⟩
  · intro x x_in_s; simp [g_zero_on_s x_in_s]
  · intro x x_in_t; simp [g_one_on_t x_in_t]
  · intro x; simp [(g_bdd x).2]

lemma exists_sum_one_of_closed_nnreal {Z : Type} [TopologicalSpace Z] [NormalSpace Z]
    {s t : Set Z} (s_closed : IsClosed s) (t_closed : IsClosed t) (disj : Disjoint s t) :
    ∃ (f₁ f₂ : Z →ᵇ ℝ≥0), EqOn f₁ 1 s ∧ EqOn f₂ 1 t ∧ f₁ + f₂ = 1 := by
  obtain ⟨f, f_zero_on_s, f_one_on_t, f_bdd⟩ := exists_bounded_zero_one_of_closed_nnreal s_closed t_closed disj
  refine ⟨1 - f, f, ?_, ?_, ?_⟩
  · intro x x_in_s; simp [f_zero_on_s x_in_s]
  · intro x x_in_t; simp [f_one_on_t x_in_t]
  · ext x
    simp only [BoundedContinuousFunction.coe_add, BoundedContinuousFunction.coe_sub,
               BoundedContinuousFunction.coe_one, Pi.add_apply, Pi.sub_apply, Pi.one_apply, NNReal.coe_add,
               NNReal.coe_one]
    norm_cast
    exact tsub_add_cancel_of_le (f_bdd x)

variable {X : Type} [TopologicalSpace X] [CompactSpace X] [T2Space X]
variable [MeasurableSpace X] [BorelSpace X]
variable (Λ : (X →ᵇ ℝ≥0) →ₗ[ℝ≥0] ℝ≥0)

lemma rieszContentAux_union {K₁ K₂ : TopologicalSpace.Compacts X} (disj : Disjoint (K₁ : Set X) K₂) :
    rieszContentAux Λ (K₁ ⊔ K₂) = rieszContentAux Λ K₁ + rieszContentAux Λ K₂ := by
  refine le_antisymm (rieszContentAux_sup_le Λ K₁ K₂) ?_
  refine le_csInf (rieszContentAux_image_nonempty Λ (K₁ ⊔ K₂)) ?_
  intro b ⟨f, ⟨hf, Λf_eq_b⟩⟩
  have ⟨g₁, g₂, hg₁, hg₂, sum_g⟩ := exists_sum_one_of_closed_nnreal K₁.isCompact.isClosed K₂.isCompact.isClosed disj
  have f_eq_sum : f = g₁ * f + g₂ * f := by simp [← RightDistribClass.right_distrib, sum_g]
  rw [← Λf_eq_b, f_eq_sum, map_add]
  have aux₁ : ∀ x ∈ K₁, 1 ≤ (g₁ * f) x := by
    intro x x_in_K₁
    simp [hg₁ x_in_K₁, hf x <| mem_union_left _ x_in_K₁]
  have aux₂ : ∀ x ∈ K₂, 1 ≤ (g₂ * f) x := by
    intro x x_in_K₂
    simp [hg₂ x_in_K₂, hf x <| mem_union_right _ x_in_K₂]
  exact add_le_add (rieszContentAux_le Λ aux₁) (rieszContentAux_le Λ aux₂)

noncomputable def rieszContent (Λ : (X →ᵇ ℝ≥0) →ₗ[ℝ≥0] ℝ≥0) : Content X where
  toFun := rieszContentAux Λ
  mono' := fun _ _ ↦ rieszContentAux_mono Λ
  sup_disjoint' := fun _ _ disj _ _ ↦ rieszContentAux_union Λ disj
  sup_le' := rieszContentAux_sup_le Λ
```

yoh-tanimoto commented:
Nice! 
I am not sure how to proceed: I wrote myself the characterization of the measure for linear `Λ`, although it needs much refactoring (one remaining `sorry` is just a variation of Urysohn's lemma, I am planning to add it in a separate PR).
https://github.com/yoh-tanimoto/mathlib4/blob/yoh/yoh/RMK/linearRMK.lean

The bundled definition of compactly supported functions is underway #12402

kkytola commented:
> Nice! I am not sure how to proceed: I wrote myself the characterization of the measure for linear `Λ`, although it needs much refactoring (one remaining `sorry` is just a variation of Urysohn's lemma, I am planning to add it in a separate PR). https://github.com/yoh-tanimoto/mathlib4/blob/yoh/yoh/RMK/linearRMK.lean
> 
> The bundled definition of compactly supported functions is underway #12402

Yes, I'm not sure what is the best way forward either!

For probability theory purposes, since convergence in distribution was defined with bounded continuous NNReal-valued functions, I would anyway like to have the bounded continuous function version of Riesz representation theorem on compact Hausdorff spaces. This was a part of my planned Prokhorov's theorem proof, which I haven't done yet, however. Of course for locally compact spaces one needs your compactly supported continuous functions anyway! So both choices come with a trade-off.

To explain a bit, I thought on compact spaces it was worth having the bounded continuous function version, since the dual norm topology then coincides with the total variation distance on finite measures and probability measures (because bounded continuous functions readily have the sup-norm instance), whereas the weak-* topology is exactly the topology of convergence in distribution. (Ok, to get the literal dual norm, one would have to cast to real values, but the weak-* topology would work directly on the `NNReal` semiring module `BoundedContinuousFunction X NNReal`.)

So I still don't have a clear view on what is the best way to implement Riesz representation theorem in Mathlib (but I would definitely like to have it in Mathlib!), although I admit to being somewhat attached to the idea that it would work seamlessly with probability theory and this is what guided my suggestion for Jesse's implementation. However, if Riesz representation is implemented with compactly supported bounded functions, then I would just add some glue to connect it to probability theory. And maybe the locally compact case is then more straightforward (Jesse and I did not do that).

Sorry that this is not really a review of your PR, but rather a continuation from the [discussion on Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/variations.20of.20Urysohn's.20lemma/near/434005302).

Mainly, I think much of this PR can be simplified, and I wanted to at least mention that now `NNReal` values are no longer the big problem they were up until now. (They are still occasionally harder to work with than real values, but on the other hand, they are somewhat more consistent with the measure theory part of the library).

kkytola commented:
> Nice! I am not sure how to proceed: I wrote myself the characterization of the measure for linear `Λ`, although it needs much refactoring (one remaining `sorry` is just a variation of Urysohn's lemma, I am planning to add it in a separate PR). https://github.com/yoh-tanimoto/mathlib4/blob/yoh/yoh/RMK/linearRMK.lean
> 
> The bundled definition of compactly supported functions is underway #12402

Actually, since you have the RMK theorem for the general locally compact case, I now think the best approach for Mathlib is indeed to get your version to Mathlib. (Jesse and I never wrote more then the compact case.) For certain probability purposes I will still want the bounded continuous function spelling, but it should be easy (and natural for Mathlib) to derive it from general case that you have. I will try to help review your PRs about that, but I'm not very experienced with that and it might still take at least a week before I get to properly trying. Overall, it is great that you have the general RMK and that you are PRing it. Thank you!

yoh-tanimoto commented:
thank you very much for your comments!

I think, as you say, specializing the locally compact case to the compact case is easy, indeed there are `liftCompactlySupported` and `compactlySupportedContinuousMapClass.ofCompact`. (they are not instances to avoid loop? I just followed what is done for `ZeroAtInfty`)

There is another question between `Real`/`NNReal`. @sgouezel suggested that such a code duplication is unavoidable. So maybe this PR should be done in `NNReal` and I will make a separate PR with `Real`?

YaelDillies commented:
I've taken the freedom to mark this PR as depending #12266 since this is what you wrote yourself but without the `- [ ] depends on:` syntax

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -36,6 +37,14 @@ variable (Λ : C_c(X, ℝ≥0) →ₗ[ℝ≥0] ℝ≥0)
 
+section Λ_mono
+
+lemma Λ_mono (f₁ f₂ : C_c(X, ℝ≥0)) (h : f₁.1 ≤ f₂.1) : Λ f₁ ≤ Λ f₂ := by
```
Does `gcongr` eat this?
```suggestion
@[gcongr]
lemma Λ_mono (f₁ f₂ : C_c(X, ℝ≥0)) (h : f₁.1 ≤ f₂.1) : Λ f₁ ≤ Λ f₂ := by
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -219,7 +230,7 @@ lemma rieszContentAux_union {K₁ K₂ : TopologicalSpace.Compacts X}
     intro x hx
     rw [mem_support]
-    exact ne_of_gt <| lt_of_lt_of_le (zero_lt_one' ℝ≥0) (hf x hx)
+    exact Ne.symm (ne_of_lt <| lt_of_lt_of_le (zero_lt_one' ℝ≥0) (hf x hx))
```
Accident?

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -245,9 +256,146 @@ lemma rieszContentAux_union {K₁ K₂ : TopologicalSpace.Compacts X}
   sup_disjoint' := fun _ _ disj _ _ ↦ rieszContentAux_union Λ disj
   sup_le' := rieszContentAux_sup_le Λ
+
+lemma rieszContent_neq_top {K : Compacts X} : rieszContent Λ K ≠ ⊤ := by
```
```suggestion
lemma rieszContent_ne_top {K : Compacts X} : rieszContent Λ K ≠ ⊤ := by
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -245,9 +256,146 @@ lemma rieszContentAux_union {K₁ K₂ : TopologicalSpace.Compacts X}
+lemma rieszContent_neq_top {K : Compacts X} : rieszContent Λ K ≠ ⊤ := by
+  simp only [ne_eq, ENNReal.coe_ne_top, not_false_eq_true]
+
+lemma rieszContentRegular : (rieszContent Λ).ContentRegular := by
```
```suggestion
lemma contentRegular_rieszContent : (rieszContent Λ).ContentRegular := by
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -245,9 +256,146 @@ lemma rieszContentAux_union {K₁ K₂ : TopologicalSpace.Compacts X}
+    rw [← NNReal.coe_mk ε (le_of_lt hε), ← NNReal.coe_add, NNReal.coe_le_coe]
+    obtain ⟨f, hfleoneonK, hfle⟩ := exists_lt_rieszContentAux_add_pos Λ K (Real.toNNReal_pos.mpr hε)
+    rw [rieszContentAux, εnneq] at hfle
+    apply le_of_lt (lt_of_le_of_lt _ hfle)
```
```suggestion
    apply le_trans _ hfle.le
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -245,9 +256,146 @@ lemma rieszContentAux_union {K₁ K₂ : TopologicalSpace.Compacts X}
+    obtain ⟨f, hfleoneonK, hfle⟩ := exists_lt_rieszContentAux_add_pos Λ K (Real.toNNReal_pos.mpr hε)
+    rw [rieszContentAux, εnneq] at hfle
+    apply le_of_lt (lt_of_le_of_lt _ hfle)
+    rw [← NNReal.coe_le_coe]
+    apply (le_iff_forall_one_lt_le_mul₀ _).mpr
```
```suggestion
    rw [← NNReal.coe_le_coe, le_iff_forall_one_lt_le_mul₀]
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -245,9 +256,146 @@ lemma rieszContentAux_union {K₁ K₂ : TopologicalSpace.Compacts X}
+    rw [← NNReal.coe_le_coe]
+    apply (le_iff_forall_one_lt_le_mul₀ _).mpr
+    · intro α hα
+      have : (Λ f) * α = Λ (α.toNNReal • f) := by
```
```suggestion
      have : Λ f * α = Λ (α.toNNReal • f) := by
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -245,9 +256,146 @@ lemma rieszContentAux_union {K₁ K₂ : TopologicalSpace.Compacts X}
+variable [T2Space X] [LocallyCompactSpace X] [MeasurableSpace X] [BorelSpace X]
+
+/-- `rieszContent` is promoted to a measure. -/
+def μ := (MeasureTheory.Content.measure (rieszContent Λ))
```
This is too ambiguous as a name. What about
```suggestion
def rieszMeasure := (MeasureTheory.Content.measure (rieszContent Λ))
```
?

YaelDillies replied:
```suggestion
def μ := (rieszContent Λ).measure
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/SetIntegral.lean:
```diff
@@ -1072,6 +1072,20 @@ theorem Continuous.integral_pos_of_hasCompactSupport_nonneg_nonzero [IsFiniteMea
 
+section Support
+
+theorem MeasureTheory.integral_tsupport [MeasurableSpace X] [TopologicalSpace X]
+    {M : Type*} [NormedAddCommGroup M] [NormedSpace ℝ M] {F : X → M} {ν : MeasureTheory.Measure X} :
+    ∫ (x : X), F x ∂ν = ∫ (x : X) in tsupport F, F x ∂ν := by
```
```suggestion
theorem MeasureTheory.integral_tsupport [MeasurableSpace X] [TopologicalSpace X]
    {M : Type*} [NormedAddCommGroup M] [NormedSpace ℝ M] {F : X → M} {ν : MeasureTheory.Measure X} :
    ∫ x in tsupport F, F x ∂ν = ∫ x, F x ∂ν := by
```

YaelDillies commented on Mathlib/Topology/ContinuousMap/CompactlySupported.lean:
```diff
@@ -541,6 +542,125 @@ noncomputable def nnrealPart (f : C_c(α, ℝ)) : C_c(α, ℝ≥0) where
+  simp only [ContinuousMap.coe_mk, Function.comp_apply, Real.coe_toNNReal', Pi.neg_apply, coe_sub,
+    coe_mk, Pi.sub_apply, max_zero_sub_max_neg_zero_eq_self]
+
+lemma toNNRalLinear_eq_iff {Λ₁ Λ₂ : C_c(α, ℝ) →ₗ[ℝ] ℝ} (hΛ₁ : ∀ f, 0 ≤ f.1 → 0 ≤ Λ₁ f)
```
```suggestion
lemma toNNRealLinear_eq_iff {Λ₁ Λ₂ : C_c(α, ℝ) →ₗ[ℝ] ℝ} (hΛ₁ : ∀ f, 0 ≤ f.1 → 0 ≤ Λ₁ f)
```

YaelDillies reviewed (commented):
Is there any chance you can split this PR further?

yoh-tanimoto commented:
The part of `rieszContentRegular` is already split #20040. Probably I can split the contents of different files separately, if that is more appropriate.

YaelDillies commented:
Ah great, I was mistaken by the lack of `- [ ] depends on:`

yoh-tanimoto commented:
probably I should define a bundled class of positive linear functionals?

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -370,12 +370,12 @@ section RieszMeasure
-    (h : tsupport f ⊆ K) : ENNReal.ofReal (Λ f) ≤ (μ Λ) K := by
-  simp only [μ, MeasureTheory.Content.measure_eq_content_of_regular (rieszContent Λ)
-    (rieszContentRegular Λ)]
+    (h : tsupport f ⊆ K) : ENNReal.ofReal (Λ f) ≤ (rieszMeasure Λ) K := by
```
```suggestion
    (h : tsupport f ⊆ K) : (Λ f).ofReal ≤ rieszMeasure Λ K := by
```

yoh-tanimoto replied:
sorry, I don't remember why I used `ofReal`, while `Λ f` is `NNReal`. I use `ofNNReal`. The dot notation does not work

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -391,7 +391,7 @@ lemma leRieszMeasure_Compacts {f : C_c(X, ℝ≥0)} (hf : ∀ (x : X), f x ≤ 1
 lemma leRieszMeasure_Opens {f : C_c(X, ℝ≥0)} (hf : ∀ (x : X), f x ≤ 1) {V : Opens X}
     (h : tsupport f ⊆ V) :
-    ENNReal.ofReal (Λ f) ≤ (μ Λ) V := by
+    ENNReal.ofReal (Λ f) ≤ (rieszMeasure Λ) V := by
```
```suggestion
    (Λ f).ofReal ≤ rieszMeasure Λ V := by
```

yoh-tanimoto replied:
same as above

YaelDillies commented on Mathlib/MeasureTheory/Integral/linearRMK.lean:
```diff
@@ -51,25 +51,25 @@ lemma mono (f₁ f₂ : C_c(X, ℝ)) (h : f₁.1 ≤ f₂.1) : Λ f₁ ≤ Λ f
+`Λ f ≤ rieszMeasure K`. -/
 lemma leRieszMeasure_of_Compacts {f : C_c(X, ℝ)} (hf : ∀ (x : X), 0 ≤ f x ∧ f x ≤ 1)
-    {K : Compacts X} (h : tsupport f ⊆ K) : ENNReal.ofReal (Λ f) ≤ (μ Λ hΛ) K := by
+    {K : Compacts X} (h : tsupport f ⊆ K) : ENNReal.ofReal (Λ f) ≤ (rieszMeasure Λ hΛ) K := by
```
```suggestion
    {K : Compacts X} (h : tsupport f ⊆ K) : (Λ f).ofReal ≤ rieszMeasure Λ hΛ K := by
```

yoh-tanimoto replied:
here `ofReal` is correct, but the dot notation does not work

YaelDillies commented on Mathlib/MeasureTheory/Integral/linearRMK.lean:
```diff
@@ -542,8 +545,8 @@ theorem RMK [Nonempty X] : ∀ (f : C_c(X, ℝ)), ∫ (x : X), f x ∂(μ Λ hΛ
-    have hΛgnleμVn : ∀ (n : Fin (⌈N⌉₊ + 1)),
-        ENNReal.ofReal (Λ (⟨g n, hg.2.2.2 n⟩)) ≤ (μ Λ hΛ) (V n) := by
+    have hΛgnlerieszMeasureVn : ∀ (n : Fin (⌈N⌉₊ + 1)),
+        ENNReal.ofReal (Λ (⟨g n, hg.2.2.2 n⟩)) ≤ (rieszMeasure Λ hΛ) (V n) := by
```
```suggestion
        (Λ (⟨g n, hg.2.2.2 n⟩)).ofReal ≤ rieszMeasure Λ hΛ (V n) := by
```

yoh-tanimoto replied:
same as above

YaelDillies commented on Mathlib/MeasureTheory/Integral/linearRMK.lean:
```diff
@@ -631,23 +638,24 @@ theorem RMK [Nonempty X] : ∀ (f : C_c(X, ℝ)), ∫ (x : X), f x ∂(μ Λ hΛ
-    simp_rw [mul_add _ ((μ Λ hΛ) _).toReal _]
+      mul_le_mul_of_nonneg_left (le_trans (hΛgnlerieszMeasureVn' n)
+        (hrieszMeasureVnlerieszMeasureEnaddε' n)) (le_of_lt (hyna n))))) _)
+    simp_rw [mul_add _ ((rieszMeasure Λ hΛ) _).toReal _]
```
```suggestion
    simp_rw [mul_add _ (rieszMeasure Λ hΛ _).toReal _]
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/linearRMK.lean:
```diff
@@ -682,20 +690,20 @@ theorem RMK [Nonempty X] : ∀ (f : C_c(X, ℝ)), ∫ (x : X), f x ∂(μ Λ hΛ
-  · calc ∫ (x : X), f x ∂(μ Λ hΛ) = ∫ (x : X), -(-f) x ∂(μ Λ hΛ) := by simp only
-      [CompactlySupportedContinuousMap.coe_neg, Pi.neg_apply, neg_neg]
-    _ = - ∫ (x : X), (-f) x ∂(μ Λ hΛ) := by exact MeasureTheory.integral_neg' (-f)
+  · calc ∫ (x : X), f x ∂(rieszMeasure Λ hΛ) = ∫ (x : X), -(-f) x ∂(rieszMeasure Λ hΛ)
+      := by simp only [CompactlySupportedContinuousMap.coe_neg, Pi.neg_apply, neg_neg]
```
```suggestion
  · calc ∫ (x : X), f x ∂(rieszMeasure Λ hΛ) = ∫ (x : X), -(-f) x ∂(rieszMeasure Λ hΛ) := by
      simp only [CompactlySupportedContinuousMap.coe_neg, Pi.neg_apply, neg_neg]
```

YaelDillies commented:
Whoops sorry, I commented on the wrong PR again 🙈 Would you mind applying the comments in the right places?

yoh-tanimoto commented:
I implemented changes to `RieszMarkovKakutani.lean` first in #20040 then merged them to this PR.

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -231,7 +231,7 @@ lemma rieszContentAux_union {K₁ K₂ : TopologicalSpace.Compacts X}
     intro x hx
     rw [mem_support]
-    exact ne_of_gt <| lt_of_lt_of_le (zero_lt_one' ℝ≥0) (hf x hx)
+    exact Ne.symm (ne_of_lt <| lt_of_lt_of_le (zero_lt_one' ℝ≥0) (hf x hx))
```
Merge accident?

yoh-tanimoto replied:
maybe, restored

YaelDillies commented on Mathlib/MeasureTheory/Integral/RealRMK.lean:
```diff
@@ -0,0 +1,709 @@
+T2 space `X` for `Real`-linear functionals `Λ`.
+
+The measure is first defined through `rieszContent` for `toNNRealLinear`-version of `Λ`.
+The result is first proved for `Real`-linear Λ because in a standard proof one has to prove the
```
```suggestion
The result is first proved for `Real`-linear `Λ` because in a standard proof one has to prove the
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RealRMK.lean:
```diff
@@ -0,0 +1,709 @@
+  MeasureTheory
+
+variable {X : Type*} [TopologicalSpace X]
+variable (Λ : C_c(X, ℝ) →ₗ[ℝ] ℝ) (hΛ : ∀ f, 0 ≤ f.1 → 0 ≤ Λ f)
```
```suggestion
variable (Λ : C_c(X, ℝ) →ₗ[ℝ] ℝ) (hΛ : ∀ f, 0 ≤ f → 0 ≤ Λ f)
```
Same below

YaelDillies commented on Mathlib/MeasureTheory/Integral/RealRMK.lean:
```diff
@@ -0,0 +1,709 @@
+  simp only [Compacts.coe_mk]
+  exact subset_rfl
+
+/-- The Riesz-Markov-Kakutani theorem for positive linear functional `Λ`. -/
```
```suggestion
/-- The **Riesz-Markov-Kakutani theorem** for a positive linear functional `Λ`. -/
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RealRMK.lean:
```diff
@@ -0,0 +1,709 @@
+  exact subset_rfl
+
+/-- The Riesz-Markov-Kakutani theorem for positive linear functional `Λ`. -/
+theorem integral_rieszMeasure_eq [Nonempty X] : ∀ (f : C_c(X, ℝ)),
```
```suggestion
theorem integral_rieszMeasure [Nonempty X] : ∀ (f : C_c(X, ℝ)),
```

yoh-tanimoto commented on Mathlib/MeasureTheory/Integral/RealRMK.lean:
```diff
@@ -0,0 +1,709 @@
+
+/-- The measure induced for `Real`-linear positive functional `Λ`, defined through `toNNRealLinear`
+and the `NNReal`-version of `rieszContent`. -/
+def rieszMeasure := (rieszContent (toNNRealLinear Λ hΛ)).measure
```
yes, done

YaelDillies replied:
Can the existing one be put in the `NNRealRMK` namespace?

yoh-tanimoto replied:
done

YaelDillies commented on Mathlib/MeasureTheory/Integral/RealRMK.lean:
```diff
@@ -0,0 +1,713 @@
+
+The measure is first defined through `rieszContent` for `toNNRealLinear`-version of `Λ`.
+The result is first proved for `Real`-linear `Λ` because in a standard proof one has to prove the
+inequalities by `le_antisymm`, yet for `C_c(X, ℝ≥0)` there is no `Neg`.
```
Something like
```suggestion
inequalities by considering `Λ f` and `Λ (-f)` for all functions `f`, yet on `C_c(X, ℝ≥0)` there is no negation.
```
maybe? Right now, it's not very clear why having to use `le_antisymm` implies we need `Neg`

YaelDillies commented on Mathlib/MeasureTheory/Integral/RealRMK.lean:
```diff
@@ -0,0 +1,713 @@
+`Λ f ≤ rieszMeasure V`. -/
+lemma le_rieszMeasure_of_isOpen_tsupport_subset {f : C_c(X, ℝ)} (hf : ∀ (x : X), 0 ≤ f x ∧ f x ≤ 1)
+    {V : Set X} (h : tsupport f ⊆ V) :
+    ENNReal.ofReal (Λ f) ≤ (rieszMeasure Λ hΛ) V := by
```
```suggestion
    .ofReal (Λ f) ≤ rieszMeasure Λ hΛ V := by
```
and you have plenty more useless brackets around `rieszMeasure Λ hΛ`

yoh-tanimoto replied:
sorry, I tend to think that `rieszMeasure Λ hΛ` is a single object. I removed brackets where possible

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani.lean:
```diff
@@ -317,6 +317,8 @@ lemma contentRegular_rieszContent : (rieszContent Λ).ContentRegular := by
 
+namespace NNRealRMK
+
 section RieszMeasure
```
Maybe you can remove this section now?

yoh-tanimoto replied:
sure, done

sgouezel commented:
Do you think it would be possible to cut the main proof into several sublemmas that would highlight the different steps in the proof? In general, it's easier to read and to maintain this way.

Also, could you update the PR description to match the current content?

YaelDillies commented:
I'm personally a bit confused about where each part of the proof goes. I would have expected the real and nnreal versions of `rieszMeasure` to be in the same file. Any reason they're not?

yoh-tanimoto commented:
I will try splitting the proof into some parts, maybe those getting `E`, `V`. But I'm not sure how to split the part doing actual estimates. Do you have any suggestion?

`RealRMK.rieszMeasure` can be in the file `RieszMarkovKakutani`, no particular reason why it is in `RealRMK`. yet the proof of the RMK theorem must be in different files in the current implementation, because first I prove the `Real`-version, then I need to import it for the `NNReal`-version, say `NNRealRMK`. Should I maybe make a folder `RieszMarkovKakutani`, and the current file will be `Basic`?

YaelDillies commented:
Yes, a folder would make sense!

yoh-tanimoto commented:
I see, should I maybe open a PR of just moving files? because otherwise diff doesn't show the changes

YaelDillies commented:
Yes, that would great, thanks!

yoh-tanimoto commented:
ok #21572

yoh-tanimoto commented:
I'm thinking about introducing structures that depend only on `f : C_c(X, ℝ)` like
```lean4
structure boundLU (f : C_c(X, ℝ)) [Nonempty X] where
/-- a lower bound -/
  a : ℝ 
/-- an upper bound -/
  b : ℝ
/-- the range -/
  L : Set ℝ
  hLdef : L = Set.range f
  hLCompact : IsCompact L
  hLNonempty : Nonempty L
  ha : a ∈ lowerBounds L
  hb : b ∈ upperBounds L
  hab : a ≤ b
  habnonneg : 0 ≤ |a| + b
```
and similarly a structure `cutRange`, giving `y` in the current code, and `cutTSupport`, giving `Erest`, and including all the lemmas in the structure. Will these be useful?

If so, I would like to know how to refer to the objects in a structure inside a proof in the structure. For example, to have an object of `boundLU`, one has to prove `hab` and would like to use `ha` `hb`, but I get something like `ha : ∀ (self : boundLU ?m.39063), self.a ∈ lowerBounds L` and not `a ∈ lowerBounds L`.

YaelDillies commented:
I am not sure I understand. `hab` can always be proved from `ha`, `hb` and `hLnonempty`. So why include it as a field in the structure? It can be a theorem instead.

yoh-tanimoto commented:
Thanks for your comments!

I thought it would be better to include lemmas in the structure just because in this way, if we have `f : C_c(X, ℝ)` `bf : boundLU f`, then `bf.hab` gives the necessary proof, which is maybe shorter than, say, `lowerBound_le_upperBound f : bf.a ≤ bf.b`?

YaelDillies commented:
> I thought it would be better to include lemmas in the structure just because in this way, if we have `f : C_c(X, ℝ)` `bf : boundLU f`, then `bf.hab` gives the necessary proof

1. That's what dot notation is for: There's no difference between having `hab` as a field of `boundLU` and proving `boundLU.hab` after the fact.
2. `hab` is a bad name for a structure field and it should be changed anyway.

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,763 @@
+lemma le_rieszMeasure_of_isOpen_tsupport_subset {f : C_c(X, ℝ)} (hf : ∀ (x : X), 0 ≤ f x ∧ f x ≤ 1)
+    {V : Set X} (h : tsupport f ⊆ V) :
+    .ofReal (Λ f) ≤ rieszMeasure Λ hΛ V := by
+  apply le_trans _ (MeasureTheory.measure_mono h)
```
```suggestion
  apply le_trans _ (measure_mono h)
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,763 @@
+  /-- the cutting point is nondecreasing. -/
+  mono' : ∀ (m n : Fin (⌈N⌉₊ + 1)), m ≤ n → val m ≤ val n
+  /-- the zeroth point is less than `a`. -/
+  zerolea : val 0 < a
```
```suggestion
  a_pos : val 0 < a
```
but I don't hugely like this because `a` is a free variable

yoh-tanimoto replied:
ok... sorry, the PR is not yet ready for review. 

Do you think it is a good idea to make a structure for `a` `b` etc. or should I just separate lemmas that are `have` in the current huge proof?

YaelDillies replied:
Can you explain this part of the proof informally?

yoh-tanimoto replied:
ok, we would like to approximate the integral of `f` over `rieszMeasure` using `Λ`. The idea is cut the support of `f` into small pieces on each of which the value of `f` is close to a certain value. Here, I want to cut the range of `f` into small pieces, which I call `y` in the current proof.

`a` is a lower bound of `range f`, `b` is an upper bound of `range f`. I pick a positive number `δ` and set `N = (b-a)/δ`. I take `y : ℕ → ℝ` in such a way that `y 0 < a < y 1 < y 2 < ... < y ⌈N⌉₊ = b` and `y (m+1) - y m < δ`.

The propositions around this part are the properties of `y`, such as monotonicity, `y 0 < a`, etc.

YaelDillies replied:
Ah I see. Then yes I think the `structure` approach makes sense

yoh-tanimoto replied:
so should I put a default value such as  `a := -‖f‖` `b ;= ‖f‖` and include the properties in the structure, or should they be theorems?

YaelDillies replied:
You should put as little as possible in the structure. But do whatever you want first, and I can give you more precise feedback afterwards.

oliver-butterley commented:
> Do you think it would be possible to cut the main proof into several sublemmas that would highlight the different steps in the proof? In general, it's easier to read and to maintain this way.

After discussing with Yoh I have completely restructured the proof and it is now divided into several separate statements. As part of the process several simplifications could be made and the entire proof is shorter. 

The heart of the main argument is presented as a calc which makes the mathematics behind the argument more easily understandable.

yoh-tanimoto commented:
I am here just to comment that Oliver greatly simplified the codes and made them readable. As he pointed out, writing big calc block and postponing the justification is really nice. I think his codes perfectly supersede mine.

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,527 @@
+
+/-- If `f` assumes values between `0` and `1` and the support is contained in `V`, then
+`Λ f ≤ rieszMeasure V`. -/
+lemma le_rieszMeasure_of_isOpen_tsupport_subset {f : C_c(X, ℝ)} (hf : ∀ (x : X), 0 ≤ f x ∧ f x ≤ 1)
```
This lemma does not assume that `V` is open, so it should rather be name `le_rieszMeasure_tsupport_subset`.

oliver-butterley replied:
Yep, unexpected `isOpen` removed

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,527 @@
+/-- If `f` assumes the value `1` on a compact set `K` then `rieszMeasure K ≤ Λ f`.-/
+lemma rieszMeasure_le_of_eq_one {f : C_c(X, ℝ)} (hf : ∀ x, 0 ≤ f x) {K : Set X}
+    (hK : IsCompact K) (hfK : ∀ x ∈ K, f x = 1) :
+    (rieszContent (toNNRealLinear Λ hΛ)).measure K ≤ ENNReal.ofReal (Λ f) := by
```
Why don't you use `rieszMeasure` on the left hand side?

oliver-butterley replied:
no reason... Updated to `rieszMeasure`,

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,527 @@
+/-- Given `f : C_c(X, ℝ)` such that `range f ⊆ [a, b]` we obtain a partition of the support of `f`
+  determined by partitioning `[a, b]` into `N` pieces. -/
+lemma range_cut_partition (f : C_c(X, ℝ)) (a : ℝ) {ε : ℝ} (hε : 0 < ε) {N : ℕ}
+    (hN : 0 < N) (hf : range f ⊆ Ioo a (a + N * ε)): ∃ (E : Fin N → Set X), tsupport f = ⋃ j, E j ∧
```
```suggestion
    (hN : 0 < N) (hf : range f ⊆ Ioo a (a + N * ε)) : ∃ (E : Fin N → Set X), tsupport f = ⋃ j, E j ∧
```

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,527 @@
+    ((ContinuousMap.measurable f.1) measurableSet_Ioc) measurableSet_closure⟩
+
+omit [LocallyCompactSpace X] in
+/-- Given a set `E`, a function `f : C_c(X, ℝ)` and `0 < ε` and `∀ x ∈ E, f x < c`, there exists a
+  set `V` such that `E ⊆ V` and the sets are similar in measure and `∀ x ∈ V, f x < c`. -/
```
specify in the docstring that `V` is open, otherwise one could take `E = V`?

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,527 @@
+/-- Choose `N` sufficiently large such that particular quantity is small. -/
+lemma RMK_le_aux {a b c ε : ℝ} (hab : a < b) (h' : 0 ≤ c) (hε : 0 < ε) : ∃ (N : ℕ), 0 < N ∧
+    (b - a) / N * (2 * c + |a| + b + (b - a) / N) ≤ ε := by
+  let ε''' := ε / (2 * c + |a| + b + 1)
```
Oops, I'm afraid this has been quite painful. Here is a streamlined way, using analysis instead of explicit computations (just saying that this tends to 0, so it has to be smaller than epsilon for large N):
```lean
  have A : Tendsto (fun (N : ℝ) ↦ (b - a) / N * (2 * c + |a| + b + (b - a) / N)) atTop
      (𝓝 (0 * (2 * c + |a| + b + 0))) := by
    apply Tendsto.mul
    · exact Tendsto.div_atTop tendsto_const_nhds tendsto_id
    · exact Tendsto.add tendsto_const_nhds (Tendsto.div_atTop tendsto_const_nhds tendsto_id)
  have B := A.comp tendsto_natCast_atTop_atTop
  simp only [add_zero, zero_mul] at B
  obtain ⟨N, hN, h'N⟩ := (((tendsto_order.1 B).2 _ hε ).and (Ici_mem_atTop 1)).exists
  exact ⟨N, h'N, hN.le⟩ 
```
(you will need to `open Filter Topology`)

oliver-butterley replied:
I feel a bit foolish having pushed through the ridiculous constructive proof when this a much better way to solve the problem. Thank you! 
Changed as suggested.

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,527 @@
+  apply le_iff_forall_pos_le_add.mpr
+  intro ε hε
+  -- Choose interval `(a, b)` which contains the range of `f`.
+  have exists_ab : ∃ a b : ℝ, a < b ∧ range f ⊆ Ioo a b := by
```
You can combine the `obtain` below with this step:
```suggestion
  obtain ⟨a, b, hab⟩ : ∃ a b : ℝ, a < b ∧ range f ⊆ Ioo a b := by
```

oliver-butterley replied:
Thanks, that's a great suggestion. Adjusted as suggested.

I was using the construction with an obtain since this was a way to encapsulate a particular part of the argument and then keep the infoview relatively clean, not cluttered with the extra things introduced during the obtain and no longer required. 

Is this generally good practice for keeping proofs clean and ordered?

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,527 @@
+        linarith
+
+/-- `Λ f ≤ ∫ (x : X), f x ∂(rieszMeasure hΛ)` -/
+theorem RMK_le [Nonempty X] (f : C_c(X, ℝ)) : Λ f ≤ ∫ (x : X), f x ∂(rieszMeasure hΛ) := by
```
I don't see why you would need `X` to be nonempty here.

oliver-butterley replied:
Good point! We are working on adjusting the argument so that it also works without requiring `[Nonempty X]`

oliver-butterley replied:
statements and proofs updated in a way that they no longer require `Nonempty X`.

oliver-butterley replied:
Thanks to Yoh for explaining to me about functions defined on empty types!

sgouezel reviewed (commented):
Thanks a lot for the reorganization of the proof, it's indeed much more readable now! Here is a first round of comments.

oliver-butterley commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,527 @@
+variable {Λ : C_c(X, ℝ) →ₗ[ℝ] ℝ} (hΛ : ∀ f, 0 ≤ f → 0 ≤ Λ f)
+
+namespace RealRMK
+
+section PositiveLinear
+
+/-- The positivity of `Λ` implies that `Λ` is monotone. -/
+lemma monotone_of_nonneg (hΛ : ∀ f, 0 ≤ f → 0 ≤ Λ f) : Monotone Λ := by
+  intro f₁ f₂ h
+  have : 0 ≤ Λ (f₂ - f₁) := by
+    apply hΛ
+    intro x
+    simp only [coe_zero, Pi.zero_apply, coe_sub, Pi.sub_apply, sub_nonneg]
+    exact h x
+  calc
+    _ ≤ Λ f₁ + Λ (f₂ - f₁) := (le_add_iff_nonneg_right (Λ f₁)).mpr this
+    _ = Λ (f₁ + (f₂ - f₁)) := Eq.symm (LinearMap.map_add Λ f₁ (f₂ - f₁))
+    _ = _ := by congr; exact add_sub_cancel f₁ f₂
+
+end PositiveLinear
```
Ah! I didn't realise there was a copy there. I'll remove this one and update the uses of it.

oliver-butterley commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,527 @@
+
+end PositiveLinear
+
+section RealRMK
```
Yes! `section RealRMK` removed

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Basic.lean:
```diff
@@ -160,7 +173,7 @@ lemma exists_continuous_add_one_of_isCompact_nnreal
     have h : x ∈ s₀ᶜ ∨ x ∈ s₁ᶜ := by
       rw [← mem_union, disj]
-      exact mem_univ _
+      trivial
```
What's the point of this change?

yoh-tanimoto replied:
I think I did this, but I don't remember why... It's ok to revert this change if explicit `exact` is better

oliver-butterley replied:
:thinking: It was changed 3 months ago so I have no idea! 

@yoh-tanimoto ?

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,469 @@
+/-
+Copyright (c) 2024 Yoh Tanimioto. All rights reserved.
```
```suggestion
Copyright (c) 2024 Yoh Tanimoto. All rights reserved.
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,469 @@
+
+/-- The measure induced for `Real`-linear positive functional `Λ`, defined through `toNNRealLinear`
+  and the `NNReal`-version of `rieszContent`. This is under the namespace `RealRMK`, while
+  `rieszMeasure` without namespace is for `NNReal`-linear `Λ`. -/
```
Please don't indent docstrings by default. Same below

oliver-butterley replied:
All changed

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,469 @@
+  apply ENNReal.coe_le_iff.mpr
+  intro p hp
+  rw [← ENNReal.ofReal_coe_nnreal, ENNReal.ofReal_eq_ofReal_iff (hΛ f hf) NNReal.zero_le_coe] at hp
+  apply csInf_le
+  · use 0
+    rw [mem_lowerBounds]
+    simp
```
```suggestion
  apply csInf_le'
```

oliver-butterley replied:
Oh! nice. thanks

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,469 @@
+  · use 0
+    rw [mem_lowerBounds]
+    simp
+  rw [Set.mem_image]
```
```suggestion
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,469 @@
+  let y : Fin N → ℝ := fun n => a + ε * (n + 1)
+  -- By definition `y n` and `y m` are separated by at least `ε`.
+  have hy {n m : Fin N} (h : n < m) : y n + ε ≤ y m := calc
+    _ ≤ a + ε * (m) + ε := by
```
```suggestion
    _ ≤ a + ε * m + ε := by
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,469 @@
+      · exact hx
+    · have (n : Fin N) : E n ⊆ tsupport f := by simp [inter_subset_right, E]
+      exact iUnion_subset this
+  exact ⟨partition, disjoint, fun n x a ↦ bdd n x a, fun _ ↦ MeasurableSet.inter
+    ((ContinuousMap.measurable f.1) measurableSet_Ioc) measurableSet_closure⟩
```
You are not using dot notation to the fullest
```suggestion
  exact ⟨partition, disjoint, fun n x a ↦ bdd n x a, fun _ ↦
    (f.1.measurable measurableSet_Ioc).inter measurableSet_closure⟩
```

oliver-butterley replied:
much better! thanks

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,469 @@
+  obtain ⟨V₁ : Opens X, hV₁⟩ := Content.outerMeasure_exists_open μ hμ hε'
+  let V₂ : Opens X := ⟨(f ⁻¹' Iio c), IsOpen.preimage f.1.2 isOpen_Iio⟩
+  use V₁ ⊓ V₂
+  have h : (∀ x ∈ V₁ ⊓ V₂, f x < c) := by
```
```suggestion
  have h x (hx : x ∈ V₁ ⊓ V₂) : f x < c := by
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,469 @@
+  exact ⟨subset_inter hV₁.1 hfE, h, h'⟩
+
+/-- Choose `N` sufficiently large such that a particular quantity is small. -/
+lemma RMK_le_aux (a' b' : ℝ) {ε : ℝ} (hε : 0 < ε) : ∃ (N : ℕ), 0 < N ∧
```
```suggestion
private lemma RMK_le_aux (a' b' : ℝ) {ε : ℝ} (hε : 0 < ε) : ∃ (N : ℕ), 0 < N ∧
```
since it's an auxiliary lemma

oliver-butterley commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/RealRMK.lean:
```diff
@@ -0,0 +1,527 @@
+  apply le_rieszMeasure_of_isCompact_tsupport_subset hΛ hf
+  · simp only [Compacts.coe_mk]
+    exact f.hasCompactSupport
+  · exact subset_rfl
```
great! modified as suggested. very efficient!

oliver-butterley commented:
Thanks Sebastien and Yael for all the comments and massive improvements! 

What generally is best for arranging a file like this? 

- Should  `RMK_le` also be marked as private? Once we have the final proof of equality we can use that rather than an inequality. Similarly with several of the other lemmas at the beginning.
- Also `iUnion_Fin_Ioc` is a mathematically obvious result that was needed for the main proof. I believe this should be marked as private. 

As far as I can see, from the outside of this file, the only relevant result is the final one, `integral_rieszMeasure`. So everything else should be marked as private?

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,436 @@
+  apply NNReal.eq
+  rw [toNNRealLinear_apply, show f.nnrealPart.toReal = f by ext z; simp [hf z], hp]
+
+/-- An `Ioc` partitions into a finite union of `Ioc`s. -/
```
Here is a simpler proof of (a variant of) this lemma, with weaker assumptions (N could be zero, and delta also, and it works in nonnegative reals, or the integers):
```lean
/-- An `Ioc` partitions into a finite union of `Ioc`s. -/
private lemma RMK_iUnion_Ioc {X : Type*} [LinearOrderedSemiring X]
    {N : ℕ} (c : X) {δ : X} (hδ : 0 ≤ δ) :
    ⋃ n ∈ Finset.range N, Ioc (c + n * δ) (c + n * δ + δ) = Ioc c (c + N * δ) := by
  induction N with
  | zero => simp
  | succ N ih => 
    simp only [Finset.mem_insert, iUnion_iUnion_eq_or_left, Nat.cast_add,
      Nat.cast_one, ih, Finset.range_succ]
    rw [union_comm, Ioc_union_Ioc_eq_Ioc]
    · simp [add_mul, add_assoc]
    · simp only [le_add_iff_nonneg_right]
      positivity
    · simp [hδ]
```
Also, I think this shouldn't be private, as this kind of decomposition is pretty common. And it shouldn't be in this file if it's not private.

oliver-butterley replied:
Thanks so much for this and all the other feedback.  

For the moment I've moved this lemma to a separate temporary file (until we find a good location for it) and have updated the main proof in order to use it. The current version is:
```
lemma iUnion_Ioc_Ioc {X : Type*} [LinearOrderedSemiring X]
    (N : ℕ) (c : X) {δ : X} (hδ : 0 ≤ δ) :
    ⋃ n ∈ Finset.range N, Ioc (c + n * δ) (c + n * δ + δ) = Ioc c (c + N * δ) := by
  induction N with
  | zero => simp
  | succ N ih =>
    simp only [Finset.mem_insert, iUnion_iUnion_eq_or_left, Nat.cast_add,
      Nat.cast_one, ih, Finset.range_succ]
    rw [union_comm, Ioc_union_Ioc_eq_Ioc]
    · simp [add_mul, add_assoc]
    · simpa [le_add_iff_nonneg_right] using mul_nonneg (Nat.cast_nonneg' N) hδ
    · simp [hδ]
```
- Is `Finset.range N` instead of `Fin N` the canonical choice?
- I made `N` explicit since it was best for my use but maybe you had chosen `N` implicit for a particular reason?
- When thinking of a general version I was naturally inclined to have a statement with a monotone function determining the end points of the interval but this became messy since there is always one more end point than interval and I didn't find a way to write it neatly. Would it make sense to push this idea? (I'm imagining analysis arguments which use exponentially shrinking intervals.) Alternatively one could take the point of view of using the uniform partition and then composing with a function when required. But then the main lemma should just be the case where `c = 0` and `δ = 1`. Any advice?
- I assume it makes sense to have the natural `Ioc`, `Ico`, `Ioi`, `Ici` alternatives or best to stick with just what we need here?

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,436 @@
+  let V₂ : Opens X := ⟨(f ⁻¹' Iio c), IsOpen.preimage f.1.2 isOpen_Iio⟩
+  use V₁ ⊓ V₂
+  have h x (hx : x ∈ V₁ ⊓ V₂) : f x < c := by
+    suffices (∀ x ∈ V₂.carrier, f x < c) by exact this x (mem_of_mem_inter_right hx)
```
```suggestion
    suffices ∀ x ∈ V₂.carrier, f x < c from this x (mem_of_mem_inter_right hx)
```

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,436 @@
+  have h x (hx : x ∈ V₁ ⊓ V₂) : f x < c := by
+    suffices (∀ x ∈ V₂.carrier, f x < c) by exact this x (mem_of_mem_inter_right hx)
+    intro _ hx
+    rw [mem_preimage, mem_Iio] at hx
```
```suggestion
```

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,436 @@
+/-- Given `f : C_c(X, ℝ)` such that `range f ⊆ [a, b]` we obtain a partition of the support of `f`
+determined by partitioning `[a, b]` into `N` pieces. -/
+private lemma RMK_range_cut (f : C_c(X, ℝ)) (a : ℝ) {ε : ℝ} (hε : 0 < ε) {N : ℕ}
+    (hN : 0 < N) (hf : range f ⊆ Ioo a (a + N * ε)) : ∃ (E : Fin N → Set X), tsupport f = ⋃ j, E j ∧
```
In this lemma and the next ones, you always have an assumption that `N` is positive, but I fail to see where this would be relevant. Can you just remove them from all the lemmas?

oliver-butterley replied:
Indeed! I removed the requirement of positive `N` in all the lemmas, adjusting them a little when required.

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,436 @@
+
+omit [T2Space X] [LocallyCompactSpace X] [MeasurableSpace X] [BorelSpace X] in
+/- Choose an interval `(a, b)` which contains the range of `f`. -/
+private lemma RMK_range (f : C_c(X, ℝ)) (hX : Nonempty X) :
```
Just remove hX from this lemma, it's not needed. For instance, you can prove it as
```lean
  obtain ⟨r, hr⟩ := (Metric.isCompact_iff_isClosed_bounded.mp 
    (HasCompactSupport.isCompact_range f.2 f.1.2)).2.subset_ball_lt 0 0
  exact ⟨-r, r, by linarith, hr.2.trans_eq (by simp [Real.ball_eq_Ioo])⟩ 
```
Since the lemma becomes essentially a one-linter, I'm not sure it's worth having as a separate lemma. If you want to keep it as a lemma, it shouldn't be private, and it should be in a better file.

oliver-butterley replied:
Nice! I've put it within the main body of the proof now it is so short

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,436 @@
+      obtain ⟨g', hg⟩ := exists_continuous_sum_one_of_isOpen_isCompact (fun n => (V n).2) f.2 this
+      exact ⟨fun n ↦ ⟨g' n, hg.2.2.2 n⟩, hg⟩
+    -- The proof is completed by a chain of inequalities.
+    calc
```
```suggestion
    calc Λ f
```
This helps the reader remember what we are proving.

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,436 @@
+      _ = Λ (∑ n, g n • f) := ?_
+      _ = ∑ n, Λ (g n • f) := by simp
+      _ ≤ ∑ n, Λ ((y n + ε') • g n) := ?_
+      _ ≤ ∑ n, (y n + ε') * Λ (g n) := by simp
```
Maybe add  a comment here that `y n + ε'` is maybe negative, which is bad for inequalities, so you force `|a|` in the expression to get positivity.

oliver-butterley replied:
Good point!

sgouezel commented:
> What generally is best for arranging a file like this?

The current organization is pretty good.

> In the file `Real.lean` there are currently 8 statements in addition to the main result.
> 
> Three of them could potentially be useful somewhere. Leave these as they are?
> 
>     * `RealRMK.le_rieszMeasure_of_isCompact_tsupport_subset`
> 
>     * `RealRMK.le_rieszMeasure_tsupport_subset`
> 
>     * `RealRMK.rieszMeasure_le_of_eq_one`

You could make `le_rieszMeasure_of_isCompact_tsupport_subset` private, as it's strictly weaker than the next lemma (there is just one more assumption). I think I would just remove this lemma and prove the strong version directly. The other two lemmas are perfect as they are: technical lemmas which are used in the proof of the main theorem, but are not an obvious corollary of it.

 
> One is an obvious result about unions of intervals that was required in the proof. Make private?
> 
>     * `iUnion_Fin_Ioc`

Rather generalize and move to a better file, as I indicated in the comments.


> Two contain arguments concerning the construction of partitions and particular coverings by open sets which are used in the proof. Make private?
> 
>     * `range_cut_partition`
> 
>     * `open_approx`

I wouldn't make them private. Who knows what someone might need some day. 

> One is the argument for the existence of sufficiently large `N`. Is set as private.
> 
>     * `RMK_le_aux`

Yes, purely technical, keep it private.

> 
> 
> The main step is an equality which is crucial in the proof but, if anyone needs to use it it can be derived in one line from the main result, it is just `le_of_eq <| Eq.symm <| RealRMK.integral_rieszMeasure hΛ f`. Make private?
> 
>     * `RMK_le`
> 
> 
> Within the main body of `RMK_le_aux` there are two parts of the argument which are contained within `obtain`s:
> 
>     * `obtain ⟨a, b, hab⟩ : ∃ a b : ℝ, a < b ∧ range f ⊆ Ioo a b` (12 lines)
> 
>     * `obtain ⟨V, hV⟩ : ∃ V : Fin N → Opens X, ∀ n, E n ⊆ (V n) ∧ (∀ x ∈ V n, f x < y n + ε') ∧ μ (V n) ≤ μ (E n) + ENNReal.ofReal (ε' / N)` (16 lines)
> 
> 
> These could be extracted and become separate private lemmas. Would that be better?

One additional comment: many lemmas have `RMK` in their name. You should remove it, because it's already implied by the namespace. On the other hand, you can maybe make some lemma names a little bit longer (and more explanatory)

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,424 @@
+      _ = ∑ n, Λ (g n • f) := by simp
+      _ ≤ ∑ n, Λ ((y n + ε') • g n) := ?_
+      _ ≤ ∑ n, (y n + ε') * Λ (g n) := by simp
+      -- That `y n + ε'` can be negative is bad the inequalities, so we artifically include `|a|`.
```
missing word in `bad the inequalities`?

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,424 @@
+      _ = Λ (∑ n, g n • f) := ?_
+      _ = ∑ n, Λ (g n • f) := by simp
+      _ ≤ ∑ n, Λ ((y n + ε') • g n) := ?_
+      _ ≤ ∑ n, (y n + ε') * Λ (g n) := by simp
```
should this be an equal sign?

oliver-butterley replied:
An equal sign in which line?

oliver-butterley replied:
Ah, I see what you mean. One moment and I will check

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,386 @@
+  apply monotone_of_nonneg hΛ
+  intro x
+  by_cases hx : x ∈ tsupport f
+  · rw [coe_toRealLinearMap, toReal_apply]
+    refine le_trans (hf x).2 (hg.1 x (mem_of_subset_of_mem (fun ⦃a⦄ a ↦ a) hx))
```
```suggestion
  · simpa using le_trans (hf x).2 (hg.1 x hx)
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,386 @@
+  let E : Fin N → Set X := fun n => (f ⁻¹' Ioc (y n - ε) (y n)) ∩ (tsupport f)
+  use E
+  -- Upper and lower bound on `f x` follow from the definition of `E n` .
+  have bdd (n : Fin N) : ∀ x ∈ E n, a + ε * n < f x ∧ f x ≤ a + ε * (n + 1) := by
+    intro x hx
```
```suggestion
  have bdd (n : Fin N) x (hx : x ∈ E n) : a + ε * n < f x ∧ f x ≤ a + ε * (n + 1) := by
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,386 @@
+    simp_rw [mem_setOf_eq, not_and_or, not_lt, not_le, or_assoc]
+    rcases (by omega : m < n ∨ n < m) with hc | hc
+    · left;
+      exact le_trans  hx.2.1 (le_tsub_of_add_le_right (hy hc))
```
```suggestion
      exact le_trans hx.2.1 (le_tsub_of_add_le_right (hy hc))
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,386 @@
+    rw [RMK_iUnion_Ioc a hε, mem_Ioc]
+    exact ⟨(hf hz).1, le_of_lt (hf hz).2⟩
+  have partition : tsupport f = ⋃ j, E j := by
+    apply subset_antisymm
+    · intro x hx
+      simp_rw [E, mem_iUnion, mem_inter_iff, mem_preimage, exists_and_right]
+      obtain ⟨j, hj⟩ := mem_iUnion.mp <| mem_of_subset_of_mem partition_aux (mem_range_self x)
+      constructor
+      · use j
+      · exact hx
+    · exact iUnion_subset (show ∀ n, E n ⊆ tsupport f by simp [inter_subset_right, E])
```
```suggestion
    simp only [E, ← iUnion_inter, ← preimage_iUnion, eq_comm (a := tsupport _), inter_eq_right]
    exact fun x hx ↦ partition_aux (mem_range_self x)
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,386 @@
+    open_approx f (div_pos hε (Nat.cast_pos'.mpr hN)) (E n) (h' n) (hν' n) (h n)
+  use V
+  intro n
+  let hV := Classical.choose_spec <|
+    open_approx f (div_pos hε (Nat.cast_pos'.mpr hN)) (E n) (h' n) (hν' n) (h n)
+  exact ⟨hV.1, hV.2.1, hV.2.2⟩
```
```suggestion
  exact Classical.choose_spec <|
    open_approx f (div_pos hε (Nat.cast_pos'.mpr hN)) (E n) (h' n) (hν' n) (h n)
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,386 @@
+/-- Choose `N` sufficiently large such that a particular quantity is small. -/
+private lemma exists_nat (a' b' : ℝ) {ε : ℝ} (hε : 0 < ε) : ∃ (N : ℕ), 0 < N ∧
+    a' / N * (b' + a' / N) ≤ ε := by
+  have A : Tendsto (fun (N : ℝ) ↦ a' / N * (b' + a' / N)) atTop (𝓝 (0 * (b' + 0))) := by
+    apply Tendsto.mul
+    · exact Tendsto.div_atTop tendsto_const_nhds tendsto_id
+    · exact Tendsto.add tendsto_const_nhds (Tendsto.div_atTop tendsto_const_nhds tendsto_id)
```
Sounds like we should make `fun_prop` handle `Tendsto` goals

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,386 @@
+integral of `f` with respect to the `rieszMeasure` associated to `L`. -/
+private lemma integral_riesz_le (f : C_c(X, ℝ)) : Λ f ≤ ∫ (x : X), f x ∂(rieszMeasure hΛ) := by
+  by_cases hX : IsEmpty X
+  -- The case `IsEmpty X` is elementry.
```
```suggestion
  -- The case `IsEmpty X` is elementary.
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,386 @@
+    let y : Fin N → ℝ := fun n ↦ a + ε' * (n + 1)
+    -- The measure of each `E n` is finite.
+    have hE' (n : Fin N) : μ (E n) ≠ ⊤ := by
+      have h : E n ⊆ tsupport f := by rw [hE.1]; exact subset_iUnion_of_subset n fun ⦃a⦄ a ↦ a
```
```suggestion
      have h : E n ⊆ tsupport f := by rw [hE.1]; exact subset_iUnion _ _
```

YaelDillies commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/iUnion_Ixx_Ixx.lean:
```diff
@@ -0,0 +1,57 @@
+lemma Fin_to_Nat {X : Type*} (N : ℕ) (s : ℕ → Set X) :
+    ⋃ (n : Fin N), s n = ⋃ n ∈ Finset.range N, s n := by
+  ext x
+  constructor
+  · intro h
+    apply Set.mem_iUnion.mpr
+    rw [Set.mem_iUnion] at h
+    obtain ⟨n, hn⟩ := h
+    use n
+    simp only [Finset.mem_range, Fin.is_lt, iUnion_true]
+    exact hn
+  · intro h
+    apply Set.mem_iUnion.mpr
+    rw [Set.mem_iUnion₂] at h
+    obtain ⟨n, hn⟩ := h
+    simp only [Finset.mem_range, exists_prop] at hn
+    use ⟨n, hn.1⟩
+    simp only
+    exact hn.2
```
```suggestion
  simp only [mem_iUnion, Finset.mem_range, exists_prop]
  constructor
  · rintro ⟨i, hi⟩
    exact ⟨i, i.2, hi⟩
  · rintro ⟨i, hiN, hi⟩
    exact ⟨⟨i, hiN⟩, hi⟩
```

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,361 @@
+    · right; left
+      exact lt_of_le_of_lt (le_tsub_of_add_le_right (hy hc)) hx.1
+  -- The sets `E n` are a partition of the support of `f`.
+  have partition_aux: range f ⊆ ⋃ n, Ioc (y n - ε) (y n) := calc
```
```suggestion
  have partition_aux : range f ⊆ ⋃ n, Ioc (y n - ε) (y n) := calc
```

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/iUnion_Ixx_Ixx.lean:
```diff
@@ -0,0 +1,42 @@
+open Set
+variable {α : Type*}
+
+lemma iUnion_Ioc_Ioc {X : Type*} [LinearOrderedSemiring X]
```
Am I correct that for your application you only need the inclusion of the right hand side in the left hand side? If so, there is a more general statement, which doesn't need multiplication or whatever, which would say: `Ico (a 0) (a N)` is contained in the union of `Ico (a i) (a (i+1))` over `i` in `range N` (and which can be trivially proved by induction over `N`). Could you check if switching to this variant works for you? (this reminds me of the very handy `sum_integral_adjacent_intervals_Ico`). This more general lemma would be named `iUnion_Ioc_subset_Ioc`, and it's definitely worth also giving the `Ico` and `Icc` versions.

By the way, this answers your question why `range N` is better than `Fin N` here: induction over `N` is very easy for the former, and not really possible for he latter.

oliver-butterley replied:
Yes, you are completely correct. Nice observation. For our present purposes it suffices to have the result:
```
lemma iUnion_Ioc_subset_Ioc {X : Type*} [LinearOrder X] (N : ℕ) (a : ℕ → X) :
    Ioc (a 0) (a N) ⊆ ⋃ i ∈ Finset.range N, Ioc (a i) (a (i + 1)) := by
```
I've added a proof of this lemma and updated `Real.lean` to use it and all is fine. (I still need to improve the lines where it is used.) I love that this lemma barely requires anything, not even assumptions on the sequence `a` which determines the end points. 

I'm sure I can improve a bit the proof of the new lemma compared to my current version but I can't see how it can be "trivially proved by induction over `N`", I see no way to avoid using cases for the different possibilities concerning how the end points are ordered. I'll think a bit more and see if something comes to mind.

sgouezel replied:
To make the proof trivial, you need to use the result for two intervals, which should be named `Ioc_subset_Ioc_union_Ioc`. I don't know if it's already in mathlib, but if it's not it's definitely worth proving separately.

oliver-butterley replied:
Ah! `Ioc_subset_Ioc_union_Ioc` is in mathlib but I failed to find it. Oops. Thanks for pointing it out, I'll update the proof using it.

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,353 @@
+    have partition_aux : range f ⊆ ⋃ n, Ioc (y n - ε) (y n) := calc
+      _ ⊆ Ioc (a + (0 : ℕ) * ε) (a + N * ε) := by
+        intro _ hz
+        simpa using (Ioo_subset_Ioc_self (hf hz))
```
```suggestion
        simpa using Ioo_subset_Ioc_self (hf hz)
```

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,353 @@
+/-- The main estimate in the proof of the Riesz-Markov-Kakutani: `Λ f` is bounded above by the
+integral of `f` with respect to the `rieszMeasure` associated to `Λ`. -/
+private lemma integral_riesz_aux (f : C_c(X, ℝ)) : Λ f ≤ ∫ x, f x ∂(rieszMeasure hΛ) := by
+  by_cases hX : IsEmpty X
+  · have : Λ f = 0 := by rw [show f = 0 by ext x; refine isEmptyElim x, LinearMap.map_zero Λ]
+    rw [integral_of_isEmpty, this]
```
I don't think you are using any more nonemptiness. So you should remove these lines, and unindent the rest of the proof.

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,353 @@
+    -- Introduce notation for the partition of the range.
+    let y : Fin N → ℝ := fun n ↦ a + ε' * (n + 1)
+    -- The measure of each `E n` is finite.
+    have hE' (n : Fin N) : μ (E n) ≠ ⊤ := by
```
```suggestion
    have hE' (n : Fin N) : μ (E n) ≠ ∞ := by
```
This is nicer, but you will need to `open scoped ENNReal`. Same thing at another place in the file.

oliver-butterley replied:
changed in both places

sgouezel commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,353 @@
+      have h n x (hx : x ∈ E n) := lt_add_of_le_of_pos ((hE.2.2.1 n x hx).right) hε'.1
+      have h' n := Eq.trans_ne
+        (Content.measure_apply (rieszContent (toNNRealLinear Λ hΛ)) (hE.2.2.2 n)).symm (hE' n)
+      use fun n ↦ Classical.choose <| exists_open_approx f h_ε' (E n) (h' n) (hE.2.2.2 n) (h n)
+      intro n
+      exact Classical.choose_spec <| exists_open_approx f h_ε' (E n) (h' n) (hE.2.2.2 n) (h n)
```
```suggestion
      choose V hV using fun n ↦ exists_open_approx f h_ε' (E n) (h' n) (hE.2.2.2 n) (h n)
      exact ⟨V, hV⟩ 
```

sgouezel commented on Mathlib/Order/Interval/Set/Union.lean:
```diff
@@ -0,0 +1,34 @@
+
+open Set
+
+/-- A union of intervals contains the interval defined by choosing any two of the end points. -/
```
```suggestion
/-- A union of consecutive intervals contains the interval defined by the initial and final points. -/
```
I am not sure what `any two of the end points` is supposed to mean. Same thing below.

sgouezel commented on Mathlib/Order/Interval/Set/Union.lean:
```diff
@@ -0,0 +1,34 @@
+/-!
+# Extra lemmas about unions of intervals
+
+This file contains lemmas about finite unions of intervals which can't be inclued with the lemmas
```
```suggestion
This file contains lemmas about finite unions of intervals which can't be included with the lemmas
```

sgouezel commented on Mathlib/Order/Interval/Set/Union.lean:
```diff
@@ -0,0 +1,34 @@
+open Set
+
+/-- A union of intervals contains the interval defined by choosing any two of the end points. -/
+theorem iUnion_Ioc_subset_Ioc {X : Type*} [LinearOrder X] (N : ℕ) (a : ℕ → X) :
```
The name is not right, it should be `Ioc_subset_biUnion_Ioc`. I'm afraid I suggested the wrong name myself :-)

Same thing for the next lemma.

sgouezel reviewed (commented):
Almost there!

oliver-butterley commented:
Oops. I should have remembered to remove the cases `IsEmpty` part of the argument after the other parts were updated. Now corrected. Sorry for overly taking advantage of the reviewer here!

oliver-butterley commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,347 @@
+          Integrable.integrableOn <| Continuous.integrable_of_hasCompactSupport f.1.2 f.2
+      _ = ∫ x in tsupport f, f x ∂μ := by simp_rw [hE.1]
+      _ = _ := setIntegral_tsupport
+  · -- Rough bound of the sum
```
This is a 20 line argument for something which is mathematically really simple. I feel there should be a way to make it more concise but failed to so better than this. Any suggestions for ways to do this better?

kkytola commented on Mathlib/MeasureTheory/Integral/RieszMarkovKakutani/Real.lean:
```diff
@@ -0,0 +1,347 @@
+          Integrable.integrableOn <| Continuous.integrable_of_hasCompactSupport f.1.2 f.2
+      _ = ∫ x in tsupport f, f x ∂μ := by simp_rw [hE.1]
+      _ = _ := setIntegral_tsupport
+  · -- Rough bound of the sum
+    rw [mul_comm 2 ε', show ε' / N = ε' * 1 / N by rw [mul_one], mul_assoc, mul_div_assoc,
+      mul_assoc, add_assoc, ← mul_add]
+    simp_rw [add_assoc |a|, add_comm (y _) ε', ← add_assoc]
+    rw [Finset.sum_add_distrib, Finset.sum_const, Finset.card_univ, Fintype.card_fin, smul_add,
+      nsmul_eq_mul, nsmul_eq_mul, ← mul_add, mul_add (1 / _), mul_comm (1 / _), mul_one_div,
+      mul_div_cancel_left₀ _ (Nat.cast_ne_zero.mpr <| Nat.not_eq_zero_of_lt hN), add_assoc _ ε',
+      show 2 * (μ K).toReal + |a| + b + ε' = 2 * (μ K).toReal + |a| + ε' + b by linarith,
+      ← add_assoc, ← add_assoc]
+    gcongr
+    · exact le_of_lt hε'.1
+    · have h : ∑ n : Fin N, y n ≤ N * b := by
+        have (n : Fin N) := calc y n
+          _ ≤ a + ε' * N := by simp_all [y, show (n : ℝ) + 1 ≤ N by norm_cast; omega]
+          _ = b := by field_simp [ε', ← mul_div_assoc, mul_div_cancel_left₀]
+        have : ∑ n, y n ≤ ∑ n, b := Finset.sum_le_sum (fun n ↦ fun _ ↦ this n)
+        simp_all
+      calc
+        _ ≤ 1 / N * (N * b) := by simp [h, hN]
+        _ ≤ _ := by simp [div_mul_cancel₀, Nat.cast_ne_zero.mpr <| Nat.not_eq_zero_of_lt hN]
```
```suggestion
  · -- Rough bound of the sum
    have h : ∑ n : Fin N, y n ≤ N * b := by
      have (n : Fin N) := calc y n
        _ ≤ a + ε' * N := by simp_all [y, show (n : ℝ) + 1 ≤ N by norm_cast; omega]
        _ = b := by field_simp [ε', ← mul_div_assoc, mul_div_cancel_left₀]
      have : ∑ n, y n ≤ ∑ n, b := Finset.sum_le_sum (fun n ↦ fun _ ↦ this n)
      simp_all 
    simp only [add_assoc, add_le_add_iff_left, Finset.sum_add_distrib, Finset.sum_add_distrib,
               Fin.sum_const, Fin.sum_const, nsmul_eq_mul, ← add_assoc, mul_add, ← mul_assoc]
    simpa [show (N : ℝ) ≠ 0 by simp [hN.ne.symm], mul_comm _ ε', div_eq_mul_inv, mul_assoc]
      using (mul_le_mul_iff_of_pos_left hε'.1).mpr <| (inv_mul_le_iff₀ (Nat.cast_pos'.mpr hN)).mpr h
```

oliver-butterley replied:
That's a huge improvement! I'll make a mental not for the future that sometimes rearranging to use gcongr is efficient but other times better to prepare the key steps and then use simp.

sgouezel commented:
bors r+
Many many thanks for getting this into mathlib, both to @yoh-tanimoto and @oliver-butterley!

oliver-butterley commented:
The bors build fails since `Order.Interval.Set.Basic` was split in #23556. I'll merge in master and then update the imports in `Union.lean`.

RemyDegenne commented:
bors r+
